src/main/scala/io/iohk/ethereum/blockchain/sync/fast/FastSync.scala

```diff
@@ -346,13 +346,13 @@ class FastSync(
         }
       }
 
-      syncState = syncState.enqueueHeaderRange(request)
       error match {
         // These are the reasons that make the master peer suspicious
         case InvalidPenultimateHeader(_, header) => handleMasterPeerFailure(header)
         case InvalidBatchHash(_, header) => handleMasterPeerFailure(header)
         // Otherwise probably it's just this peer's fault
         case _ =>
+          syncState = syncState.enqueueHeaderRange(request)
```

Comment:
The current failed request (for headers) is not required to be enqueued if the master peer is invalid. We are discarding those and a fresh request is created based on the result obtained from BranchResolver. Hence this change.